### PR TITLE
User, role, userRole controllers

### PR DESCRIPTION
**Pull request by skosinskiy**

### CODE REVIEW

**Comment by skosinskiy** [comment]
[3h - Create endpoints for User, Role, UserRole](https://trello.com/c/Z1aiblim/116-3h-create-endpoints-for-user-role-userrole)

File: src/main/java/com/danit/finalproject/application/repository/UserRepository.java [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+import com.danit.finalproject.application.entity.User;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Repository;
+
+import java.util.List;
+
+@Repository
+public interface UserRepository extends JpaRepository<User, Long> {
+
+  List<User> findAllByEmailStartingWith(String email);
```
Comment (zozich) [inline src/main/java/com/danit/finalproject/application/repository/UserRepository.java]: Let's also add `ignoreCase` here

File: src/main/java/com/danit/finalproject/application/service/RoleService.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+  public Role updateRole(Long roleId, Role role) {
+    role.setId(roleId);
+    return roleRepository.save(role);
+  }
+
+  public Role deleteRole(Long roleId) {
+    Role role = roleRepository.findById(roleId).orElse(null);
+    if (role != null) {
+      role.getUsers().forEach(user -> user.removeRole(role));
```
Comment (zozich) [inline src/main/java/com/danit/finalproject/application/service/RoleService.java]: That's bad practice to have action methods inside entities.
1. You should do it here in service
2. You were not saving the updated users

File: src/main/java/com/danit/finalproject/application/controller/UserController.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+import java.util.List;
+
+@RestController
+@RequestMapping("api/users")
+public class UserController {
+
+  @Autowired
+  private UserService userService;
+
+  @GetMapping("{userId}")
```
Comment (zozich) [inline src/main/java/com/danit/finalproject/application/controller/UserController.java]: Where will we use this endpoint?

File: src/main/java/com/danit/finalproject/application/entity/Role.java [hunk 4]
```diff
@@ -19,10 +20,21 @@
...
+  @ManyToMany(mappedBy = "roles")
+  private List<User> users;
+
   @ManyToMany
   @JoinTable(name = "permissions_roles",
           joinColumns = {@JoinColumn(name = "role_id")},
           inverseJoinColumns = {@JoinColumn(name = "permission_id")})
   private List<Permission> permissions;
 
+  @Override
```
Comment (zozich) [inline src/main/java/com/danit/finalproject/application/entity/Role.java]: Why do you need that? Lombok's `@Data` annotation provides also `toString()` implementations. Use annotations to control it's behavior

File: src/main/java/com/danit/finalproject/application/entity/User.java [hunk 5]
```diff
@@ -37,7 +37,7 @@
   private Long age;
 
   @Column(name = "gender")
-  private boolean gender;
+  private Gender gender;
```
Comment (zozich) [inline src/main/java/com/danit/finalproject/application/entity/User.java]: :+1:

File: src/main/java/com/danit/finalproject/application/entity/User.java [hunk 6]
```diff
@@ -61,4 +61,8 @@
   @JoinColumn(name = "visit_id")
   private List<Visit> visits;
 
+  public void removeRole(Role role) {
```
Comment (zozich) [inline src/main/java/com/danit/finalproject/application/entity/User.java]: Bad practice

File: src/main/java/com/danit/finalproject/application/controller/RoleController.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+import java.util.List;
+
+@RestController
+@RequestMapping("api/roles")
+public class RoleController {
+
+  @Autowired
```
Comment (zozich) [inline src/main/java/com/danit/finalproject/application/controller/RoleController.java]: That's bad practice to use autowiring. Please use constructor autowiring instead - change in the rest of files as well

File: src/main/resources/application.yml [hunk 8]
```diff
@@ -0,0 +1,6 @@
+spring:
```
Comment (zozich) [inline src/main/resources/application.yml]: Let's add here 9000 port as well

File: src/test/java/com/danit/finalproject/application/service/RoleServiceTest.java [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.MOCK)
+@AutoConfigureMockMvc
+@Transactional
+public class RoleServiceTest {
+
+  @Autowired
+  private RoleService roleService;
+
+  @Test
+  public void getRoleById() {
```
Comment (zozich) [inline src/test/java/com/danit/finalproject/application/service/RoleServiceTest.java]: 1. We agreed to test service layer with unit tests. For it to work this way you need to mock repository layer and do not make any real DB calls
2. This is a bad name for unit test - you should specify what this test is doing (not that critical for controller tests above, but also can be changed)

File: src/main/java/com/danit/finalproject/application/controller/UserController.java (hunk 3 above)
Comment (javasia) [reply src/main/java/com/danit/finalproject/application/controller/UserController.java]: @zozich But it's stated in the REST API architecture, isn't it?

File: src/main/resources/application.yml (hunk 8 above)
Comment (skosinskiy) [reply src/main/resources/application.yml]: I have defined another port for tests just for possibility of running tests when app is running.

File: src/main/java/com/danit/finalproject/application/controller/UserController.java (hunk 3 above)
Comment (zozich) [reply src/main/java/com/danit/finalproject/application/controller/UserController.java]: Ok, NVM then

**Pull request merged**